jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererCodeCoverageTracker.kt

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.codewhisperer.telemetry
+
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.editor.event.DocumentEvent
+import com.intellij.util.Alarm
+import com.intellij.util.AlarmFactory
+import org.jetbrains.annotations.TestOnly
+import software.aws.toolkits.jetbrains.services.codewhisperer.model.InvocationContext
+import software.aws.toolkits.jetbrains.services.codewhisperer.model.SessionContext
+import software.aws.toolkits.jetbrains.services.codewhisperer.popup.CodeWhispererPopupManager
+import software.aws.toolkits.jetbrains.services.codewhisperer.popup.CodeWhispererUserActionListener
+import software.aws.toolkits.jetbrains.services.codewhisperer.util.CodeWhispererConstants.TOTAL_SECONDS_IN_MINUTE
+import software.aws.toolkits.telemetry.CodewhispererLanguage
+import software.aws.toolkits.telemetry.CodewhispererTelemetry
+import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.math.roundToInt
+
+abstract class CodeWhispererCodeCoverageTracker(
+    private val timeWindowInSec: Long,
+    private val language: CodewhispererLanguage,
+    acceptedTokensSize: AtomicInteger,
+    totalTokensSize: AtomicInteger
+) : Disposable {
+    val percentage: Int
+        get() = if (totalTokensSize.get() != 0) (acceptedTokensSize.get().toDouble() / totalTokensSize.get() * 100).roundToInt() else 0
+    var acceptedTokensSize = acceptedTokensSize
+        private set
+    var totalTokensSize = totalTokensSize
+        private set
+    private val alarm = AlarmFactory.getInstance().create(Alarm.ThreadToUse.POOLED_THREAD, this)
+    private val isShuttingDown = AtomicBoolean(false)
+    private var startTime: Instant = Instant.now()
+
+    init {
+        val conn = ApplicationManager.getApplication().messageBus.connect()
+        conn.subscribe(
+            CodeWhispererPopupManager.CODEWHISPERER_USER_ACTION_PERFORMED,
+            object : CodeWhispererUserActionListener {
+                override fun afterAccept(states: InvocationContext, sessionContext: SessionContext, remainingRecomm: String) {
+                    if (states.requestContext.fileContextInfo.programmingLanguage.toCodeWhispererLanguage() != language) return
+                    addAndGetAcceptedTokens(remainingRecomm.length)
+                }
+            }
+        )
+        scheduleCodeWhispererCodeCoverageTracker()
+    }
+
+    fun documentChanged(event: DocumentEvent) {
+        addAndGetTotalTokens(event.newLength - event.oldLength)
+    }
+
+    private fun flush() {
+        try {
+            if (isTelemetryEnabled()) emitCodeWhispererCodeContribution()
+        } finally {
+            init()
+            scheduleCodeWhispererCodeCoverageTracker()
+        }
+    }
```

Comment:
I am confused by this function -- why do we call init() here?

Reply:
I see it refers to the init() function down below - so do we reset these values in every 5-min window? Is this aligned with VSC's implemetation?

Reply:
yes I should have name it as `reset`, which should make more sense on its purpose. Yes it aligns with the design doc and VSC impl.

Reply:
updated